File: OvalensPlanner/apps/api/app/models/tax.py
"""Tax calculation models."""

from pydantic import BaseModel


class TaxBand(BaseModel):
    name: str
    lower_limit: float
    upper_limit: float | None
    rate: float


class IncomeSummary(BaseModel):
    employment: float = 0
    self_employment: float = 0
    pension: float = 0
    rental: float = 0
    savings: float = 0
    dividends: float = 0
    other: float = 0

    @property
    def total_income(self) -> float:
        return (
            self.employment
            + self.self_employment
            + self.pension
            + self.rental
            + self.savings
            + self.dividends
            + self.other
        )


class TaxCalculation(BaseModel):
    taxable_income: float
    income_tax: float
    national_insurance: float
    total_tax: float
    effective_rate: float
    marginal_rate: float
    bands: list[TaxBand] = []


class AdjustedNetIncome(BaseModel):
    total_income: float
    pension_contributions: float = 0
    gift_aid: float = 0
    adjusted_net_income: float = 0

File: OvalensPlanner/packages/shared/src/types/tax.ts
export interface TaxBand {
  name: string;
  lowerLimit: number;
  upperLimit: number | null;
  rate: number;
}

export interface IncomeSummary {
  employment: number;
  selfEmployment: number;
  pension: number;
  rental: number;
  savings: number;
  dividends: number;
  other: number;
  totalIncome: number;
}

export interface TaxCalculation {
  taxableIncome: number;
  incomeTax: number;
  nationalInsurance: number;
  totalTax: number;
  effectiveRate: number;
  marginalRate: number;
  bands: TaxBand[];
}

export interface AdjustedNetIncome {
  totalIncome: number;
  pensionContributions: number;
  giftAid: number;
  adjustedNetIncome: number;
}
